src/LayerTree/LayerTree.tsx

```diff
@@ -35,618 +51,342 @@ interface OwnProps {
-        this.rebuildTreeNodes();
-      }
-    }
+    return visibleKeys;
+  }, [filterFunction, layerGroup, map]);
 
-    if (nodeTitleRenderer !== prevProps.nodeTitleRenderer) {
-      this.rebuildTreeNodes();
+  const getTreeNodeTitle = useCallback((layer: OlLayerBase) => {
+    if (_isFunction(nodeTitleRenderer)) {
+      return nodeTitleRenderer(layer);
+    } else {
+      return layer.get('name');
     }
-  }
+  }, [nodeTitleRenderer]);
 
-  /**
-   * Determines what to do when the component is unmounted.
-   */
-  componentWillUnmount() {
-    unByKey(this.olListenerKeys);
-  }
-
-  /**
-   * Creates TreeNodes from a given layergroup and sets the treeNodes in the state.
-   *
-   * @param groupLayer A grouplayer.
-   */
-  treeNodesFromLayerGroup(groupLayer: OlLayerGroup) {
-    const layerArray = groupLayer.getLayers().getArray()
-      .filter(this.props.filterFunction);
-    const treeNodes = layerArray.map((layer) => {
-      return this.treeNodeFromLayer(layer);
+  const hasListener = useCallback((key: OlEventsKey) => {
+    return olListenerKeys.current.some(listenerKey => {
+      return listenerKey.target === key.target
+        && listenerKey.type === key.type
+        && listenerKey.listener === key.listener;
     });
-    treeNodes.reverse();
-    this.setState({ treeNodes });
-  }
+  }, []);
 
-  /**
-   * Registers the add/remove listeners recursively for all ol.layer.Group.
-   *
-   * @param groupLayer A ol.layer.Group
-   */
-  registerAddRemoveListeners(groupLayer: OlLayerGroup) {
-    const collection = groupLayer.getLayers();
-    const addEvtKey = collection.on('add', this.onCollectionAdd);
-    const removeEvtKey = collection.on('remove', this.onCollectionRemove);
-    const changeEvtKey = groupLayer.on('change:layers', this.onChangeLayers);
+  const treeNodeFromLayer = useCallback((layer: OlLayerBase) => {
+    let childNodes: TreeDataNode[] = [];
 
-    this.olListenerKeys.push(addEvtKey, removeEvtKey, changeEvtKey);
+    if (filterFunction && [layer].filter(filterFunction).length === 0) {
+      return;
+    }
 
-    collection.forEach((layer) => {
-      if ((layer as OlLayerGroup).getLayers) {
-        this.registerAddRemoveListeners((layer as OlLayerGroup));
+    if (layer instanceof OlLayerGroup) {
+      let childLayers = layer.getLayers().getArray();
+      if (filterFunction) {
+        childLayers = childLayers.filter(filterFunction);
       }
-    });
-  }
-
-  /**
-   * Registers an eventhandler on the `ol.View`, which will rebuild the tree
-   * nodes whenever the view's resolution changes.
-   */
-  registerResolutionChangeHandler() {
-    const { map } = this.props;
-    const evtKey = map.on('moveend', this.rebuildTreeNodes.bind(this));
-    this.olListenerKeys.push(evtKey); // TODO when and how to we unbind?
-  }
+      childNodes = childLayers
+        .map(childLayer => treeNodeFromLayer(childLayer))
+        .filter(childLayer => childLayer !== undefined)
+        .reverse() as TreeDataNode[];
```

Comment:
not important in this case, but `toReversed` is preferred because it does not change in place. So it's good practice to always use that instead of `reverse` imo.
```suggestion
        .toReversed() as TreeDataNode[];
```

Reply:
Applied in bb9f693bf.